Turn querying mapped provider before execution to an error

eskatos opened the pull request:
See https://github.com/gradle/gradle/issues/15876

eskatos commented:
@bot-gradle test QFL

bot-gradle commented:
<!-- {"replyTargetCommentId":789569288,"teamCityBuildId":"41799854","teamCityBuildHeadRef":"775be6091008f8748f6bc57c38dfd4e62140643d","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [QuickFeedbackLinux build](https://builds.gradle.org/viewQueued.html?itemId=41799854) for you.

eskatos commented:
@bot-gradle test RFM

bot-gradle commented:
<!-- {"replyTargetCommentId":789633459,"teamCityBuildId":"41804205","teamCityBuildHeadRef":"6eafca0d13345da8d6a37d34814ede01a984f578","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [ReadyForMerge build](https://builds.gradle.org/viewQueued.html?itemId=41804205) for you.

eskatos commented:
@bot-gradle test RFN

bot-gradle commented:
<!-- {"replyTargetCommentId":789718611,"teamCityBuildId":"41808559","teamCityBuildHeadRef":"c79de90d12871a621fc53ecebf05762a06f6a97d","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [ReadyForNightly build](https://builds.gradle.org/viewQueued.html?itemId=41808559) for you.

eskatos commented:
@bot-gradle test RFN

bot-gradle commented:
<!-- {"replyTargetCommentId":789719788,"teamCityBuildId":"41808910","teamCityBuildHeadRef":"bcec290c86794fc84707d2b0547c6816002d81e2","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already cancelled the [old build](https://builds.gradle.org/viewLog.html?buildId=41808559) and triggered a new [ReadyForNightly build](https://builds.gradle.org/viewQueued.html?itemId=41808910) for you.

bamboo commented on subprojects/docs/src/docs/userguide/migration/upgrading_version_6.adoc:
```diff
@@ -258,6 +258,12 @@ checkstyle {
+==== Querying the mapped value of a provider before the producer has completed is now an error
+
+Gradle 6.x warns users about the wrong behavior and then returns a possibly incorrect provider value.
+Starting from 7.0 the same case will produce an error.
```
💄 _Starting with_ seems to be a little more popular. :)

bamboo commented on subprojects/docs/src/docs/userguide/migration/upgrading_version_6.adoc:
```diff
@@ -258,6 +258,12 @@ checkstyle {
+
+Gradle 6.x warns users about the wrong behavior and then returns a possibly incorrect provider value.
+Starting from 7.0 the same case will produce an error.
+Plugins and build scripts should be adjusted to query the mapped value of a provider, for example a task output property, after the producer has completed.
```
💄 I'd replace _producer_ by _task_ to make it more concrete and match the example.

bamboo commented on subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/ThirdPartyGradleModuleMetadataSmokeTest.groovy:
```diff
@@ -34,7 +34,7 @@ class ThirdPartyGradleModuleMetadataSmokeTest extends AbstractSmokeTest {
         // TODO Test Kotlin 1.4
         def kotlinVersion = TestedVersions.kotlin.latestStartsWith("1.3")
-        def androidPluginVersion = AGP_VERSIONS.getLatestOfMinor("4.0")
+        def androidPluginVersion = AGP_VERSIONS.getLatestOfMinor("4.1")
```
❓ I assume the version bump is related because 4.0 would now fail?

eskatos replied:
Correct, as wrote in the commit message

bamboo reviewed (approved):
LGTM!

eskatos commented:
Thanks @bamboo 
I addressed your comments and further reduced AGP coverage now that 4.0 will break.
Could you please take another look?

@bot-gradle test RFN

bot-gradle commented:
<!-- {"replyTargetCommentId":789787511,"teamCityBuildId":"41812550","teamCityBuildHeadRef":"78a67bd12e7a8de1e321a40c324bbc37ede02fd2","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [ReadyForNightly build](https://builds.gradle.org/viewQueued.html?itemId=41812550) for you.
